tests/distributor/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+    #[derive(Default)]
+    struct ProgramState {
+        nodes: BTreeSet<Program>,
+        amount: u64,
+    }
+
+    static mut STATE: Option<ProgramState> = None;
+
+    impl Program {
+        fn new(handle: impl Into<ProgramId>) -> Self {
+            Self {
+                handle: handle.into(),
+            }
+        }
+
+        fn do_request<Req: Encode, Rep: Decode>(
+            &self,
+            request: Req,
+        ) -> impl Future<Output = Result<Rep, &'static str>> {
+            let encoded_request: Vec<u8> = request.encode();
+
+            let program_handle = self.handle;
+            async move {
+                let reply_bytes = gstd_async::msg::send_and_wait_for_reply(
+                    program_handle,
+                    &encoded_request[..],
+                    msg::gas_available() - 25_000_000,
+                    0,
+                )
+                .await;
+
+                Rep::decode(&mut &reply_bytes[..]).map_err(|_| "Failed to decode reply")
+            }
+        }
+
+        async fn do_send(&self, amount: u64) -> Result<(), &'static str> {
+            match self.do_request(Request::Receive(amount)).await? {
+                Reply::Success => Ok(()),
+                _ => Err("Unexpected send reply"),
+            }
+        }
+
+        async fn do_report(&self) -> Result<u64, &'static str> {
+            match self.do_request(Request::Report).await? {
+                Reply::Amount(amount) => Ok(amount),
+                _ => Err("Unexpected send reply"),
+            }
+        }
+
+        fn nodes() -> &'static mut BTreeSet<Program> {
+            unsafe { &mut STATE.as_mut().expect("STATE UNITIALIZED!").nodes }
+        }
+
+        fn amount() -> &'static mut u64 {
+            unsafe { &mut STATE.as_mut().expect("STATE UNITIALIZED!").amount }
+        }
+
+        async fn handle_request() {
+            let reply = match msg::load::<Request>() {
+                Ok(request) => match request {
+                    Request::Receive(amount) => Self::handle_receive(amount).await,
+                    Request::Join(program_id) => Self::handle_join(program_id).await,
+                    Request::Report => Self::handle_report().await,
+                },
+                Err(e) => {
+                    ext::debug(&format!("Error processing request: {:?}", e));
+                    Reply::Failure
+                }
+            };
+
+            msg::reply(&reply.encode()[..], msg::gas_available() - 25_000_000, 0);
+        }
+
+        async fn handle_receive(amount: u64) -> Reply {
+            let subnodes_count = Self::nodes().len() as u64;
+            if subnodes_count > 0 {
+                let distributed = amount / subnodes_count * subnodes_count;
+                let mut left_over = amount - distributed;
+
+                if distributed > 0 {
+                    for program in Program::nodes().iter() {
+                        if let Err(_) = program.do_send(distributed).await {
```

Comment:
this is no longer safe to do since the node list can be modified by another program, thus refcell should be used and the `Request::Join` should fail if the node is processing sub-nodes at the moment